### PR TITLE
MaterialUserObject - on-demand recalculatable "material properties"

### PR DESCRIPTION
**Pull request by dschwen**
I want to put this idea out here. It is not quite done (needs to override the execute parameter, needs `final`, etc.)

This would allow for a more elegant design of the plasticity system, and will fix the HEVP system. I'm off to ski now.

### CODE REVIEW

**Comment by dschwen** [comment]
Yeah, I forgot that I have to make all HEVP user objects inherit from the new class :-)

**Comment by dschwen** [comment]
Sorry, I edited from my phone. I guess that hammered civet pretty bad :-)

**Comment by mangerij** [comment]
you are skiing and pushing code? are the chair lifts that boring in Idaho?
 anyway I wanted to say I am all for updating the user object system to
make it more streamlined for the end user. Plasticity makes perfect sense.
I think we will need a BCUserObject - that way a whole host of boundary
integral methods could be implemented and rewrite the MOOSE routines in
favor of some libMesh one. Ill take a look at what you are doing and maybe
try to follow along with this?

**Comment by dschwen** [comment]
If you are interested in BCUserObjects for boundary integrals you will likely need to have Jacobian terms provided by UserObjects. Check out #5914 and help me figure out why the Jacobians are incorrect.

This PR goes in a different direction. It provides a new class of pluggable functions for use in nested solved in material classes.

File: framework/include/materials/Material.h [hunk 1]
```diff
@@ -258,6 +259,25 @@ class Material :
...
   virtual QpData * createData();
 
+  /**
+   * Get a apecial MaterialUserObject which can be explicitly called by the material
+   * during the material property calculation.
+   * @param name The name of the parameter key of the user object to retrieve
+   * @return The user object with name associated with the parameter 'name'
+   */
+  template<class T>
+  const T & getMaterialUserObject(const std::string & name);
```
Comment (friedmud) [inline framework/include/materials/Material.h]: I don't think these new functions are necessary.  This should just be handled on the back end in the Warehouse.  The Warehouse should return either `tid=0` or the correct `tid` depending on what kind of UserObject is being requested.  `UserObjectInterface` can be modified to pass the `tid` down when it makes the call.

No reason to add extra "special" functions for this new type of `UserObject`

**Comment by friedmud** [comment]
Ok - I'm not against the entire idea... but I want a better name.  This could be a generally useful thing... even outside of Materials.

How about `HelperUserObject`?  Or something along those lines?  Or, since it's going to be specific to Elements still it would be an `ElementHelperUserObject`...

**Comment by friedmud** [comment]
`SubUserObject`?  `DiscreteUserObject` (because it's executed on a discrete object like an Element or a Side or whatever)?

`GeometricObjectUserObject`?

`ObjectThatYouCallFromAnotherObjectToDoSomeObjectivelyAwesomeObjectOnObjectComputation`?

**Comment by dschwen** [comment]
Haha! I'll let you settle on a name. The PR here is just an idea. It would fix the threading issues and is fairly non-invasive. The type of UO proposed here would be useful everywhere a value needs to be recomputed multiple times during a residual or Jacobian calculation, for example in a nested solve. 
I guess those could happen in Kernels just as well as in Materials, but the name comes from the idea of having a recalculatable (is that even a word?) replacement of _Material_Properties.

The interface specific to Materials (as opposed to having getUserObject magically return the "right" thing) is not a bad thing IMO. This type of UO is used quite differently from other UOs. But I can see the appeal of having a unified interface. 

It would be even cooler, if the UOs magically knew the current `_qp`. I thought about passing in a reference to the Material's `_qp` member, in my `getMaterialUserObject` method, but a UO may be gotten by multiple materials :-/

Anyhow keep the suggestions coming!

File: framework/include/base/FEProblem.h [hunk 2]
```diff
@@ -481,12 +481,12 @@ class FEProblem :
    * @return Const reference to the user object
    */
   template <class T>
-  const T & getUserObject(const std::string & name)
+  const T & getUserObject(const std::string & name, unsigned int tid = 0)
```
Comment (permcody) [inline framework/include/base/FEProblem.h]: Due to the discussion in https://github.com/libMesh/libmesh/issues/801, this PR is DOA. We cannot retrieve thread specific versions of an object and expect them to contain specific data.

Tagging @friedmud

Comment (dschwen) [inline framework/include/base/FEProblem.h]: Yeah, I guess I don't really see the reason for this, as we are staying in the **same** thread in this design. I'm also not asking for specific data, I'm merely relying on accessing the initialized element in that thread. 

Again, no cross-thread access, no internal state of the userobject that is prepared by MOOSE.

Comment (permcody) [inline framework/include/base/FEProblem.h]: If you are really inside of the same threaded loop then this is valid code. However, you are modifying the FEProblem API. So this won't work in general. It won't even work if you try to retrieve an elemental user object from a side user object where we have a different threaded loop, or better yet an elemental pre-aux user object from an elemental post-aux user-object! We need to come up with a way of doing this without opening a hole in the framework that can be accidentally exploited.

Comment (dschwen) [inline framework/include/base/FEProblem.h]: No. It should work even if those are different threaded loops. The identity of the thread is meaningless. What matters is the _tid, which is used to look up the prepared elements etc. If those are the same in both objects they will access consistent data!

Comment (permcody) [inline framework/include/base/FEProblem.h]: OK - I can see that. Just note that somebody will mess this up in the future.

**Comment by dschwen** [comment]
I see, if execute_on is empty the UO isn't even created.

File: framework/include/materials/Material.h (hunk 1 above)
Comment (permcody) [inline framework/include/materials/Material.h]: I like this idea. It'll avoid exposing `tid` to the user so people won't be able to screw up the retrieval of normal UserObjects.

**Comment by friedmud** [comment]
Wait - why was this merged?  There were several open comments... and the name of the new object isn't right either...

**Comment by dschwen** [comment]
I guess I may have been too pushy :angel: 

Anyhow, of course I just realized that for my LDRD I need to call these objects from `ElementUserObject`s :-D. We'll it is not too late to fix this.

File: framework/include/materials/Material.h (hunk 1 above)
Comment (aeslaughter) [inline framework/include/materials/Material.h]: I will add another opinion to these, getUserObject should just handle it. If a MaterialUserObject is requested, then use the tid, if it is not then don't. You are doing a dynamic_cast to MaterialUserObject in this function, so just do this in the regular getUserObject and have the cast determine what to do with the tid.

File: framework/src/userobject/MaterialUserObject.C [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+MaterialUserObject::MaterialUserObject(const InputParameters & parameters) :
+    ElementUserObject(parameters)
+{
+}
+
+void
+MaterialUserObject::execute()
+{
+  mooseError("MaterialUserObjects must be called explicitly from Materials");
```
Comment (aeslaughter) [inline framework/src/userobject/MaterialUserObject.C]: Why only materials? Anything acting on elements would work.

File: framework/include/materials/Material.h (hunk 1 above)
Comment (dschwen) [inline framework/include/materials/Material.h]: Andrew, I wrote this in a different issue, but as dynamic_cast does not work with a mere forward declare your suggestion is not trivial to implement! I may go down that route anyhow if I figure out a way....

File: framework/src/userobject/MaterialUserObject.C (hunk 3 above)
Comment (dschwen) [inline framework/src/userobject/MaterialUserObject.C]: ...because, as I commented already,I will need this from elementuserobjexts

**Comment by dschwen** [comment]
I'm working on a PR that makes `getUserObject` do the right thing automatically. Testing now.

**Pull request merged**